extensions/sql-database-projects/src/models/deploy/deployService.ts

```diff
@@ -0,0 +1,403 @@
+				if (createdDockerId) {
+					// Get the docker logs if docker was created but crashed
+					await this.executeCommand(constants.dockerLogMessage(createdDockerId));
+				}
+			}
+
+			return undefined;
+		});
+	}
+
+	private async buildAndRunDockerContainer(dockerFilePath: string, imageName: string, root: string, profile: ILocalDbSetting, imageLabel: string): Promise<string | undefined> {
+		this.logToOutput('Building docker image ...');
+		await this.executeCommand(`docker pull ${constants.dockerBaseImage}`);
+		await this.executeCommand(`docker build -f ${dockerFilePath} -t ${imageName} ${root}`);
+		await this.executeCommand(`docker images --filter label=${imageLabel}`);
+
+		this.logToOutput('Running docker container ...');
+		await this.executeCommand(`docker run -p ${profile.port}:1433 -e "MSSQL_SA_PASSWORD=${profile.password}" -d ${imageName}`);
+		return await this.executeCommand(`docker ps -q -a --filter label=${imageLabel} -q`);
+	}
+
+	private async getConnectionString(connectionUri: string): Promise<string | undefined> {
+		const getAzdataApi = await utils.getAzdataApi();
+		//const vscodeMssqlApi = getAzdataApi ? undefined : await utils.getVscodeMssqlApi();
+		if (getAzdataApi) {
+			const connection = await getAzdataApi.connection.getConnection(connectionUri);
+			if (connection) {
+				return await getAzdataApi.connection.getConnectionString(connection.connectionId, true);
+			}
+		}
+		// TODO: vscode connections string
+
+		return undefined;
+
+	}
+
+	private async connectToDatabase(profile: ILocalDbSetting, savePassword: boolean, database: string): Promise<ConnectionResult | string | undefined> {
+		const getAzdataApi = await utils.getAzdataApi();
+		const vscodeMssqlApi = getAzdataApi ? undefined : await utils.getVscodeMssqlApi();
+		if (getAzdataApi) {
+			const connectionProfile = {
+				password: profile.password,
+				serverName: `${profile.serverName},${profile.port}`,
+				database: database,
+				savePassword: savePassword,
+				userName: profile.userName,
+				providerName: 'MSSQL',
+				saveProfile: false,
+				id: '',
+				connectionName: `${constants.connectionNamePrefix} ${profile.dbName}`,
+				options: [],
+				authenticationType: 'SqlLogin'
+			};
+			return await getAzdataApi.connection.connect(connectionProfile, false, false);
+		} else if (vscodeMssqlApi) {
+			const connectionProfile = {
+				password: profile.password,
+				server: `${profile.serverName}`,
+				port: profile.port,
+				database: database,
+				savePassword: savePassword,
+				user: profile.userName,
+				authenticationType: 'SqlLogin',
+				encrypt: false,
+				connectTimeout: 30,
+				applicationName: 'SQL Database Project'
+			};
+			let connectionUrl = await vscodeMssqlApi.connect(connectionProfile);
+			return connectionUrl;
+		} else {
+			return undefined;
+		}
+	}
+
+	private async validateConnection(connection: ConnectionResult | string | undefined): Promise<ValidationResult> {
```

Comment:
Would suggest adding comments to a lot of these functions. It's not always clear what exactly they're doing - for example what does it mean to "validate" a connection?